ui/js/controllers/jobs.js

```diff
@@ -177,19 +177,63 @@ treeherderApp.controller('ResultSetCtrl', [
             $rootScope.$emit(thEvents.deleteRunnableJobs, $scope.resultset);
         };
 
-        $scope.cancelAllJobs = function(revision) {
-            if (!window.confirm('This will cancel all pending and running jobs for revision ' + revision + '!\n\nClick "OK" if you\'re sure.')) {
-                return;
+        $scope.getCancelJobsTitle = function() {
+            var job = ThResultSetStore.getSelectedJob($scope.repoName).job;
+            var singleJobSelected = job instanceof ThJobModel;
+            if (singleJobSelected) {
+                if ($scope.canCancelJobs()) {
+                    return "Cancel selected job";
+                } else {
+                    return "Cannot cancel completed job";
+                }
+            } else {
+                return "Cancel all jobs";
             }
+        };
 
-            ThResultSetModel.cancelAll($scope.resultset.id, $scope.repoName).then(function() {
-                return thBuildApi.cancelAll($scope.repoName, revision);
-            }).catch(function(e) {
-                thNotify.send(
-                    ThModelErrors.format(e, "Failed to cancel all jobs"),
-                    'danger', true
-                );
-            });
+        $scope.canCancelJobs = function() {
```

Comment:
Could you actually add something like this here:

```js
            if (!$scope.user.loggedin) {
                return "Must be logged in to cancel jobs";
            }
```

This will disable the element if the user isn't logged in, which is nicer than the present behavior. :)